### DISCUSSION TITLE
Method to filter assets within mosaic_reader

### DISCUSSION THREAD

**Discussion by sor-droneup** [cogeotiff/rio-tiler#683] [Q&A]
Hello,

I've noticed that for some mosaic the mosaic_reader function download some geotiffs before pixel selection, even if those geotiffs does not intersect with the tile (this happens in case if mosaic was generated with low zoom level). 

A simple filter method could be added that would check if geotiff boundaries intersects with a tile before loading it, that would greatly reduce the execution time in scenarios like above.

Has this been considered before ? Is there any reason why there is no such filtering?

Such logics could be added for example in line 78 here: https://github.com/cogeotiff/rio-tiler/blob/73175d0607ef28bffcb6f03ff8dfd2d6abfbc191/rio_tiler/mosaic/reader.py#L25

**Comment by vincentsarago** [comment]
if the geotiff does not intersect with the tile, a `TileOutsideBounds` https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/io/rasterio.py#L348-L351 should be raised. But in all the case, this will imply that the COG header will be downloaded and the COG dataset opened. 

could you maybe explain a bit more and share a reproducible example 🙏

**Reply by sor-droneup** [reply]
H, thanks for the answer! 
What I am thinking is including the boundary box and crs into the mosaicjson spec, so the file does need to be read in order to find out that the tile lies outside of the bounds (this could be done on mosaic reader level). I've noticed that opening a file becomes the bottle neck when the traffic spikes, since it spawns separate thread for each one. Currently i'm getting away with it by maintaining multiple mosaicjsons of the same dataset (with different quadkey zoom level), however a vision of having just one Mosaic which also includes the boundary data seem more performant. I'm wondering if solution like that had been considered?
